domain/port/state/state.go

```diff
@@ -0,0 +1,277 @@
+	}
+
+	insertPortRange, err := st.Prepare("INSERT INTO port_range (*) VALUES ($unitPortRange.*)", unitPortRange{})
+	if err != nil {
+		return errors.Annotate(err, "preparing insert port range statement")
+	}
+
+	clearPortRange, err := st.Prepare("DELETE FROM port_range WHERE unit_endpoint_uuid IN ($endpointUUIDs[:]) ", endpointUUIDs{})
+	if err != nil {
+		return errors.Annotate(err, "preparing clear port range statement")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+
+		// Get the endpoints which currently exist in state.
+		// Cache the endpoint's UUIDs for later use.
+		var currenUnitEndpoint []endpoint
+		err = tx.Query(ctx, getUnitEndpoints, unitUUID).GetAll(&currenUnitEndpoint)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			currenUnitEndpoint = []endpoint{}
+		} else if err != nil {
+			return domain.CoerceError(err)
+		}
+
+		foundEndpoints := set.NewStrings()
+		unitEndpointMap := map[string]endpointUUID{}
+		for _, ep := range currenUnitEndpoint {
+			foundEndpoints.Add(ep.Endpoint)
+			unitEndpointMap[ep.Endpoint] = endpointUUID{UUID: ep.UUID}
+		}
+
+		// Insert any new endpoints that are required.
+		requiredEndpoints := set.NewStrings(endpointsUnderAction...).Difference(foundEndpoints)
+		newUnitEndpoints := make([]unitEndpoint, requiredEndpoints.Size())
+		for i, endpoint := range requiredEndpoints.Values() {
+			uuid, err := uuid.NewUUID()
+			if err != nil {
+				return errors.Annotatef(err, "generating UUID for unit endpoint")
+			}
+			newUnitEndpoints[i] = unitEndpoint{
+				UUID:     uuid.String(),
+				UnitUUID: unit,
+				Endpoint: endpoint,
+			}
+			unitEndpointMap[endpoint] = endpointUUID{UUID: uuid.String()}
+		}
+
+		if len(newUnitEndpoints) > 0 {
+			err = tx.Query(ctx, insertUnitEndpoint, newUnitEndpoints).Run()
+			if err != nil {
+				return domain.CoerceError(err)
+			}
+		}
+
+		// Construct a map of protocol names to protocol IDs for later use.
+		protocols := []protocol{}
+		protocolMap := map[string]int{}
+		err = tx.Query(ctx, getProtocols).GetAll(&protocols)
+		if err != nil {
+			return domain.CoerceError(err)
+		}
+		for _, protocol := range protocols {
+			protocolMap[protocol.Name] = protocol.ID
+		}
+
+		// Get the current opened ports for the unit's endpoints.
+		// This will be used to calculate the changes required.
+		var currentOpenedEndpointPorts []endpointPortRange
+		err = tx.Query(ctx, getOpenedPortsOnEndpoints, unitUUID, endpointsUnderAction).GetAll(&currentOpenedEndpointPorts)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			currentOpenedEndpointPorts = []endpointPortRange{}
+		} else if err != nil {
+			return domain.CoerceError(err)
+		}
+
+		endpointMap := map[string][]portRange{}
+		for _, r := range currentOpenedEndpointPorts {
+			endpointMap[r.Endpoint] = append(endpointMap[r.Endpoint], portRange{
+				Protocol: r.Protocol,
+				FromPort: r.FromPort,
+				ToPort:   r.ToPort,
+			})
+		}
+
+		unitPortRanges := []unitPortRange{}
```

Comment:
I prefer these to be declared as `var` unless an empty non-nil slice has special meaning. It doesn't appear to here.

Reply:
I think because I append I initialised to empty slice. However, I remember now that you can append to nil, and take the length.